Feat: Player Rating Endpoints

invaderb opened the pull request:
feat: add ratings endpoints
feat: add scheduled job
test: add tests for rating generation
chore: add endpoints to readme

Adds the following endpoints to the service:
/v1/ratings
/v1/ratings/{player_name}

Changes:
Adds the following dependency @nest/schedule

nitzel committed:
chore[Readme]: make endpoints list more readable

nitzel committed:
fix: use path.join to prevent misconfigurations of e.g. DB_PATH when the ending slash is missing

invaderb committed:
feat: add rating endpoints
feat: add scheduled job
test: add tests for rating generation
chore: ad endpoints to readme

invaderb committed:
feat: add rating endpoints
feat: add scheduled job
test: add tests for rating generation
chore: ad endpoints to readme

invaderb committed:
feat: add rating endpoints
feat: add scheduled job
test: add tests for rating generation
chore: ad endpoints to readme

invaderb committed:
fix: comment typos

invaderb committed:
fix: comment typos

invaderb committed:
fix: select values
fix: add missing api param
fix: make optional columns

invaderb committed:
fix: select values
fix: add missing api param
fix: make optional columns

invaderb committed:
fix: swagger docs

invaderb committed:
fix: swagger docs

invaderb committed:
chore: remove unneeded console.log

invaderb committed:
chore: remove unneeded console.log

nitzel commented on src/app.module.ts:
```diff
@@ -1,29 +1,42 @@
 			synchronize: false,
 		}),
+		TypeOrmModule.forRoot({
+			name: 'default',
```
`BaseDataSourceOptions.name` is [deprecated](https://github.com/typeorm/typeorm/blob/d8a2e3730f12bb2b8e521635e176a284594121f3/src/data-source/BaseDataSourceOptions.ts#L23), [explanation:  `we don't need names anymore since we are going to drop all related methods relying on this property.`](https://github.com/typeorm/typeorm/blob/d8a2e3730f12bb2b8e521635e176a284594121f3/src/data-source/DataSource.ts#L62)

The alternative seems to be to tell TypeOrm here which entities are related to which database:

https://docs.nestjs.com/techniques/database#repository-pattern

```ts
TypeOrmModule.forRoot({ 
      // ...
      database: process.env.DB_PATH + 'players.db',
      entities: [Ratings, Players],
    }),
```

invaderb replied:
Interesting I wonder how it's going to work with two different db's then; I'll look into it and push a fix.

invaderb replied:
Yeah it breaks if you remove the name when managing multiple databases on both the for root and for feature.  it needs some way to differentiate between the 2 and I found that you need to have a default for the datasource to fall back on. if we had one db this wouldn't be a problem and could use this pattern.

invaderb replied:
![Screenshot 2023-05-06 at 3 18 12 PM](https://user-images.githubusercontent.com/16693116/236646942-f7c96dc5-ed64-4070-8f76-3cbca2cd3f10.jpg)

invaderb replied:
I also tried the autoLoadEntities and that didn't work either.

invaderb replied:
Ah yeah here in the docs, https://docs.nestjs.com/techniques/database#multiple-databases
```
In this case, data source naming becomes mandatory.
```

nitzel replied:
Ugh sorry for sending you down that path. I'm confused why they a) deprecated `name` if it's still necessary and b) what is `entities` for?

nitzel commented on src/models/ratings/ratings.module.ts:
```diff
@@ -0,0 +1,26 @@
+@Module({
+	controllers: [RatingsController],
+	imports: [
+		TypeOrmModule.forFeature([Ratings, Players], 'default'),
```
no need to specify connection name, see other comment. `[Ratings, Players, Games]` probably can be imported in a single `forFeature` call.

nitzel commented on src/models/ratings/ratings.service.spec.ts:
```diff
@@ -0,0 +1,321 @@
+		expect(service).toBeDefined();
+	});
+	
+	// describe('Generate Ratings', () => {	
```
?

invaderb replied:
more so just left this in here to come back and add more unit tests for the generateRating method but I'll just go ahead and remove it, as the main business logic has testing.

nitzel commented on src/models/ratings/ratings.service.ts:
```diff
@@ -0,0 +1,434 @@
+				totalPages: Math.ceil(results[1] / limit),
+			};
+		} catch (error) {
+			console.error(error);
```
why not use `this.logger.error`?

invaderb replied:
Just forgot to move it over to logger, which I don't mind using but it does strip out the line number where the error happened which makes it harder to debug...

nitzel replied:
I think the `console.log` can go the next time you need to touch this file 😄

nitzel commented on .env:
```diff
@@ -5,4 +5,5 @@ NODE_ENV=local
-DIST_PATH=dist/
\ No newline at end of file
+DIST_PATH=dist/
+PREVIOUS_FILE=previous.txt
```
What do you think about renaming this to `last_used_game_id.txt` or similar?

invaderb replied:
I'm fine either way this was just copied over from the old script.

invaderb replied:
for backwards compatibility going to leave this as is so there is parity between what's used on the server is the same as in the repo.

nitzel reviewed (commented):
LGTM - just a few minor suggestions.

Should this be pointing to the `dev` branch?

invaderb committed:
chore: code clean up and refactor
fix: add missing db names to typeorm feature and root

invaderb committed:
chore: code clean up and refactor
fix: add missing db names to typeorm feature and root

invaderb committed:
chore: code clean up

invaderb committed:
chore: code clean up

invaderb committed:
chore: code clean up

invaderb committed:
chore: code clean up

invaderb committed:
chore: code clean up

invaderb committed:
chore: code clean up

invaderb committed:
Merge branch 'dev' into rating-endpoints

invaderb committed:
test: fix game service tests

invaderb committed:
test: fix game service tests

invaderb committed:
Merge branch 'rating-endpoints' of github.com:USTakAssociation/playtak-api into rating-endpoints

invaderb committed:
Merge branch 'rating-endpoints' of github.com:USTakAssociation/playtak-api into rating-endpoints

invaderb committed:
Merge branch 'rating-endpoints' of github.com:USTakAssociation/playtak-api into rating-endpoints

invaderb committed:
Merge branch 'rating-endpoints' of github.com:USTakAssociation/playtak-api into rating-endpoints

invaderb committed:
fix: issues from rebasing

nitzel commented on src/models/ratings/dto/ratings.dto.ts:
```diff
@@ -0,0 +1,68 @@
+		description: 'Number of results per page',
+		required: false
+	})
+		limit?: string;
```
Should this be a `number`? Same with `RatingQuery .page`?

invaderb replied:
Since these come from the query string they are strings and not numbers and need to be parsed after the fact.

nitzel commented on src/models/ratings/dto/ratings.dto.ts:
```diff
@@ -0,0 +1,68 @@
+	})
+		page?: string;
+	@ApiProperty({
+		description: 'Optional skip if you dont use page',
```
what is `skip` for? What values can I pass in?

invaderb replied:
it's the total number of items to skip when doing pagination, can clean up this description

nitzel commented on src/models/ratings/ratings.controller.ts:
```diff
@@ -0,0 +1,35 @@
+	@ApiResponse({status: 500, type: DefaultExceptionDto, description: 'Returns 500 server error'})
+	@ApiParam({ name: 'name', description: 'player name', required: true })
+	@Get('/:name')
+	findRating(@Param() param: any): Promise<Rating> {
```
I think this can be improved to 
```ts
	findRating(@Param('name') name: string): Promise<Rating> {
		return this.service.getPlayersRating(name);
	}
```

invaderb committed:
chore: code clean up and refactor
chore: add better api property descriptions

humanat commented on src/models/ratings/dto/ratings.dto.ts:
```diff
@@ -29,38 +29,38 @@ export class RatingList {
 export class RatingQuery {
 	@ApiProperty({
-		description: 'Number of results per page',
+		description: 'Number of results per pag for pagination. Optional.',
```
pag = page

invaderb committed:
fix: spelling error

humanat commented on README.md:
```diff
@@ -53,6 +52,12 @@ https://api.{env}.playtak.com
+- /v1/ratings{player_name}
+
+- /v1/ratings
+- /v1/ratings{player_name}
```
Should these have a / in them before the player_name or is that baked into the variable?

invaderb replied:
yeah need to add the / I'll make an update to the readme and double check the endpoints

humanat reviewed (approved):
Approving, assuming that the comments are trivial to explain.

invaderb committed:
fix: readme endpoints

invaderb committed:
Merge branch 'dev' into rating-endpoints

invaderb committed:
Merge branch 'dev' into rating-endpoints

invaderb committed:
feat: add bru api test files

devp commented on src/models/ratings/tasks/rating.task.ts:
```diff
@@ -0,0 +1,19 @@
+import { Injectable, Logger } from '@nestjs/common';
+import { Cron } from '@nestjs/schedule';
```
TIL about node-cron. wow.

invaderb replied:
yeah it's a nifty package and pretty easy to setup in nest.

devp commented on src/main.ts:
```diff
@@ -22,11 +23,6 @@ async function bootstrap() {
 	});
 	
 	const config = new DocumentBuilder()
-		.addApiKey({
```
Was this apiKey never used?

invaderb replied:
No this was just copy paste from the example

invaderb committed:
fix: fatigue rating calculation
chore: refactor to make variables more meaningful
chore: refactor ratings to have better typings
chore: refactor dto files
chore: update formatting
chore: fix typos

invaderb committed:
fix: unit test for updated typings
chore: fix typings
chore: clean up player dto

devp commented on src/models/dto/response.dto.ts:
```diff
@@ -0,0 +1,10 @@
+/* eslint-disable indent */
+import { ApiProperty } from '@nestjs/swagger';
+
+export class DefaultReposeDto {
```
typo

devp commented on src/models/ratings/ratings.controller.ts:
```diff
@@ -0,0 +1,41 @@
+		return this.service.getPlayersRating(name);
+	}
+
+	// // manually kick off rating generation
```
This is just a TODO but this should probably be a POST

invaderb replied:
the endpoint isn't meant to be public, it's only here for testing locally why it's commented out the method type doesn't really matter in this case.

if we do ever make it public and put it behind some admin check we'll change it to a post at that point.

invaderb committed:
fix: typos
chore: update readme

invaderb committed:
chore: update readme

devp commented on src/models/ratings/ratings.service.ts:
```diff
@@ -0,0 +1,443 @@
+	}
+
+	// returns the adjusted rating of a player
+	public adjustPlayer(
```
Note that this method and updateFatigue are mutations; they change the `player` object that is passed in. The description suggests otherwise. That may be fine though?

invaderb replied:
Updated the comments to explain what its doing it better.

devp commented on src/models/ratings/ratings.service.ts:
```diff
@@ -0,0 +1,443 @@
+		}
+		return (player.fatigue[opponentID] = (player.fatigue[opponentID] || 0) + gameFactor);
+	}
+
```
contrasting to adjustPlayer, adjustedRating just returns the value

invaderb replied:
I updated the comments to explain it better what it's doing.

devp commented on src/models/ratings/ratings.service.ts:
Is the goal with ratings.service that the generateRating() code is ported from the nohat code mostly, and that we can run a command in NodeJS land rather than running a sciript?

generateRating() was too dense for much of a line-by-line reading -- its correctness will have to be proven out by running it over the real data and seeing that it's the same - but I don't know if it's worth rewriting in a future branch, either.

invaderb replied:
The current nohat script is just that, a standalone nodejs script that is currently called with a system cron job. It works however, it's kinda a mess there's no comments explaining what it's doing, logging, no unit tests, no automated deployments, or following any standards that we're now defining with the playtak-api.

The goal is is bring it into the playtak api that runs the generateRating with the nestjs schedule rather than the system cron job, including all the other updates in this PR.

lol try reading the nohat version. I'm pretty confident on this now with reviewing the unit testing again and compared generated rating data locally

devp commented on src/models/ratings/ratings.service.spec.ts:
I like the unit test, but have to vaguely trust what it is testing.

invaderb replied:
I spent a bit of time going through the nohat script line by line one what it's doing and replicated that in the unit tests.

devp reviewed (approved):
LGTM but leaving some non-blocking comments for my understanding and some things I noticed. I most notably just skimmed over generateRating() because that was ported code that is hard to parse, and even riskier to rewrite without testing.

invaderb committed:
fix: logic for always adding fatiguerating to players who have or haven't changed
chore: add additional comments to explain the generateRatings

invaderb committed:
fix: add cron directive

nitzel commented on previous.txt:
```diff
@@ -0,0 +1 @@
+547436 hash
```
Is this equivalent to selecting the oldest game entry in the `games` table that doesn't have ratings yet? Wouldn't that be preferable over using an external file to tell us where to start the rating algorithm?

invaderb replied:
Currently the Tak server always saves a completed game with a default adjusted rating then the rating script comes through and updates to what the correct adjusted rating is which is difficult to ensure we could use that to allow us to see what games have or haven't been rated. 

In the future we can make a change to the games db to start tracking if a game has actually been rated or not which would allow us to remove the need for saving the last game ID.

invaderb committed:
fix: update deploy script
